src/util/any.hpp

```diff
@@ -0,0 +1,269 @@
+    }
+};
+
+class any {
+public:
+    constexpr any() = default;
+
+    any(const any& other):
+        state_(other.state_->copy())
+    {}
+
+    any(any&& other) noexcept {
+        std::swap(other.state_, state_);
+    }
+
+    template <
+        typename T,
+        typename = typename
+            util::enable_if_t<!std::is_same<util::decay_t<T>, any>::value>
+    >
+    any(T&& other) {
+        using contained_type = util::decay_t<T>;
+        static_assert(
+            std::is_copy_constructible<contained_type>::value,
+            "Type of contained object stored in any must satisfy the CopyConstructible requirements.");
+
+        state_.reset(new model<contained_type>(std::forward<T>(other)));
+    }
+
+    any& operator=(const any& other) {
+        state_.reset(other.state_->copy());
+        return *this;
+    }
+
+    any& operator=(any&& other) noexcept {
+        swap(other);
+        return *this;
+    }
+
+    template <
+        typename T,
+        typename = typename
+            util::enable_if_t<!std::is_same<util::decay_t<T>, any>::value>
+    >
+    any& operator=(T&& other) {
+        using contained_type = util::decay_t<T>;
+
+        static_assert(
+            std::is_copy_constructible<contained_type>::value,
+            "Type of contained object stored in any must satisfy the CopyConstructible requirements.");
+
+        state_.reset(new model<contained_type>(std::forward<T>(other)));
+        return *this;
+    }
+
+    void reset() noexcept {
+        state_.reset(nullptr);
+    }
+
+    void swap(any& other) noexcept {
+        std::swap(other.state_, state_);
+    }
+
+    bool has_value() const noexcept {
+        return (bool)state_;
+    }
+
+    const std::type_info& type() const noexcept {
+        return state_->type();
+    }
+
+private:
+
+    struct interface {
+        virtual ~interface() = default;
+        virtual const std::type_info& type() = 0;
+        virtual interface* copy() = 0;
+        virtual void* pointer() = 0;
+        virtual const void* pointer() const = 0;
+    };
+
+    template <typename T>
+    struct model: public interface {
+        ~model() = default;
+
+        interface* copy() override {
+            return new model<T>(*this);
+        }
+
+        const std::type_info& type() override {
+            return typeid(T);
+        }
+
+        model(const T& other):
+            value(other)
+        {}
+
+        model(T&& other):
+            value(std::move(other))
+        {}
+
+        void* pointer() {
+            return &value;
+        }
+
+        const void* pointer() const {
+            return &value;
+        }
+
+        T value;
+    };
+
+    std::unique_ptr<interface> state_;
+
+protected:
+
+    template <typename T>
+    friend const T* any_cast(const any* operand);
+
+    template <typename T>
+    friend T* any_cast(any* operand);
+
+    template <typename T>
+    T* unsafe_cast() {
+        return reinterpret_cast<T*>(state_->pointer());
```

Comment:
You are right. This was a hang over from a slightly different earlier implementation which required a more "industrial" solution.